File: Source/Triggers/CancelDeathTrigger.cs
namespace Celeste.Mod.MintChocolateHelper.Triggers;

[Tracked]
[CustomEntity("MintChocolateHelper/CancelDeathTrigger")]
public class CancelDeathTrigger : Trigger
{
    private readonly int Delay;
    internal readonly bool DisableQuickRespawn;
    internal readonly bool SkipEverestEventOnDie;
    internal readonly bool AffectRetries;
    internal readonly bool DontRegisterDeathInStats;
    internal readonly bool KeepFollowers;

    internal readonly string Flag;
    internal readonly object FlagExpression;
    internal readonly bool IsValidExpression;

    public CancelDeathTrigger(EntityData data, Vector2 offset) : base(data, offset)
    {
        Delay = data.Int("delay");
        DisableQuickRespawn = data.Bool("disableQuickRespawn");
        SkipEverestEventOnDie = data.Bool("skipEverestEventOnDie");
        AffectRetries = data.Bool("affectRetries");
        DontRegisterDeathInStats = data.Bool("dontRegisterDeathInStats");
        KeepFollowers = data.Bool("keepFollowers");

        Flag = data.Attr("flag");
        if (FrostHelperImports.SafeTryCreateSessionExpression(Flag, out FlagExpression)) IsValidExpression = true;
    }

    public override void OnEnter(Player player)
    {
        base.OnEnter(player);
        if (MintChocolateHelperModule.Session.PlayerIsPseudoDead) Add(new Coroutine(Unkill(Delay)));
    }

    private static IEnumerator Unkill(int delay)
    {
        yield return delay / 60f;
        if (!MintChocolateHelperModule.Session.PlayerIsPseudoDead) yield break;
        MintChocolateHelperModule.Session.CancelDeathTriggerTeleportingPlayer = true;
    }
}

File: Loenn/triggers/cancel_death_trigger.lua
local CancelDeathTrigger = {}

CancelDeathTrigger.name = "MintChocolateHelper/CancelDeathTrigger"

CancelDeathTrigger.placements = {
    name = "cancel_death_trigger",
	alternativeName = {"unkill_player_trigger", "revive_player_trigger"},
    data = {
        width = 16,
        height = 16,
        flag = "",
        delay = 0,
		disableQuickRespawn = false,
		skipEverestEventDie = false,
		affectRetries = false,
        dontRegisterDeathInStats = false,
		keepFollowers = false
    }
}

CancelDeathTrigger.fieldInformation = {
    delay = {
        fieldType = "integer",
    }
}

CancelDeathTrigger.fieldOrder = {
    "x", "y",
    "width", "height",
    "flag", "delay",
    "disableQuickRespawn", "skipEverestEventDie",
	"affectRetries", "dontRegisterDeathInStats",
	"keepFollowers"
}

return CancelDeathTrigger